README.md
# linter-gcc

[![Join the chat at https://gitter.im/linter-gcc/Lobby](https://badges.gitter.im/linter-gcc/Lobby.svg)](https://gitter.im/linter-gcc/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://travis-ci.com/AtomLinter/linter-gcc.svg?branch=master)](https://travis-ci.com/AtomLinter/linter-gcc) [![apm](https://img.shields.io/apm/dm/linter-gcc.svg?style=flat-square)](https://atom.io/packages/linter-gcc)

This plugin provides an interface to gcc/g++ for linting and works with [Linter](https://github.com/AtomLinter/Linter).

Used with files with grammar "C", "C++" and ["C++14"](https://atom.io/packages/language-cpp14).

Includes linting **on-the-fly**, though it is only partially tested. Please open an issue if you encounter any problems.

## Important info for Mac OSX users!
If you have XCode installed on OSX, the `gcc/g++` commands will both link to `clang`. This can cause issues with the `-fmax-errors` option used by linter-gcc, which isn't recognised by clang. To properly install GCC, you need to install it with Homebrew (instructions [here](https://github.com/hebaishi/linter-gcc/issues/62)). You will likely also need to add the flag `-fsyntax-only` to your C Flags or C++ Flags.

## Linter in action!

![linter-gcc screenshot](https://raw.githubusercontent.com/hebaishi/images/master/lintergcc_onthefly.gif)

## Using CMake compile settings
linter-gcc can take compile settings from CMake. For example:

```bash
git clone https://github.com/hebaishi/gtf2tab
cd gtf2tab
mkdir build
cd build
cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 ..
```

Running ```cmake``` with the ```-DCMAKE_EXPORT_COMPILE_COMMANDS``` flag generates a ```compile_commands.json``` file which linter-gcc can get the compile settings from. Then you simply open the project in Atom, and enter ```./build/compile_commands.json``` in the Compile Commands File setting of linter-gcc. Note that if you supply a valid ```compile_commands.json``` file, your include paths and compile flags configuration settings (described below) are ignored.

Unfortunately, CMake does not typically compile header files so the ```compile_commands.json``` file does not include entries for these files. To add them you can use a tool like [compdb](https://github.com/Sarcasm/compdb).

```bash
compdb -p ./ list > compile_commands.json
```

## File/Project-Specific settings

Assuming you have the a file called ```sample.cpp``` open, linter-gcc performs the following actions:

1. Looks for file called ```sample.cpp.gcc-flags.json``` in the same directory as your source file (file-specific settings)
2. Looks for a file called ```.gcc-flags.json``` in every subdirectory from the current up to your project root (subdirectory/project-specific settings)
3. If no ```.gcc-flags.json``` is found, the settings in your configuration page are used.

The package takes its settings from the first configuration file that is found.

You can specify your settings in ```.gcc-flags.json```, at any level (file/subdirectory/project) using the following syntax:

```json
{
  "execPath": "/usr/bin/g++",
  "gccDefaultCFlags": "-Wall",
  "gccDefaultCppFlags": "-Wall -std=c++11",
  "gccErrorLimit": 15,
  "gccIncludePaths": ".,./include,./path",
  "gccSuppressWarnings": true
}
```

Note that the include paths need to be separated by commas. If this file is present, it will replace the settings you specified in the settings window. Relative paths (starting with ```.``` or ```..```) are expanded with respect to the root folder. Both ```execPath``` and ```gccIncludePaths``` are expanded.

Paths are expanded relative to the project that contains the file being linted. The package limits its search to 30 levels when looking for a configuration file. Within each project, you may have as many file/directory-specific configuration files as you wish.

### Usage notes:
* Add ```-fsyntax-only``` to your C/C++ compilation flags to prevent the generation of ```.gch``` files when linting headers
* Add ```-c``` to your flags to avoid linking errors.

### Plugin installation
Press ctrl and ',' or cmd and ',' , click on 'Packages', search 'linter gcc', or:
```
$ apm install linter-gcc
```
### Reporting Issues

Please read the [Wiki](https://github.com/AtomLinter/linter-gcc/wiki) before reporting any issues.


CHANGELOG.md
## 0.8.7
* Fix bug causing files to only use compile_commands.json args on first lint

## 0.8.6
* Add iquote path when using lintOnTheFly to mimic gcc include path behavior
* Add default position property to default message created when parsing fails

## 0.8.5
* Use method to read compile_commands and config files which will properly detect modifications

## 0.8.2
* Enable handling of multiple projects in one window
* Add reference to compdb and new Wiki issues section hyperlink

## 0.8.1
* Remove Linter as a hard dependency (version 0.8.0 made it optional)
* Remove atom-package-deps usage
* Replace atom-linter dependency with sb-exec

## 0.8.0
* Updated to Linter v2
* Added placemark (^~~) parsing for better highlighting if GCC 7+ is used
* Added option to re-lint other files that have contributed messages to the current file
* Cleanup messages from other files on save if using lintOnSave
* Added separate handling of isystem include paths

## 0.7.1
* Fix for 'trim' undefined error and compile_commands.json

## 0.7.0
* Add combined C/C++ compile path
* Add CMake integration
* Merge pull reguests for French language Fix
* Merge pull request for Deprecated selector

## 0.6.15
* Fix undefined string error

## 0.6.14
* Add description to max-errors option

## 0.6.13
* Fix error when modifying unsaved buffer

## 0.6.12
* Make Paypal button smaller

## 0.6.11
* Add Paypal button

## 0.6.10
* Add recursive include path expansion
* Add option to turn off ```-fmax-errors```

## 0.6.9
* Fix linter-gcc reporting incorrect filename

## 0.6.8
* Use JSON Stringify workaround for string length error
* Make temp file grammar-specific for linting on-the-fly

## 0.6.7
* Re-factor + improve grammar type validation

## 0.6.6
* Streamline + make debug messages optional
* Fix babel import error
* Re-arrange package settings
* Re-enable automatic dependency installation

## 0.6.5
* Beautify README.md

## 0.6.4
* Modify string split to accept escaped characters
* Capture compile flag-related errors

## 0.6.3
* Add lint on-the-fly animation to README.md

## 0.6.2
* Complete C++14 support

## 0.6.1
* Add C++14 support

## 0.6.0
* Add lint on-the-fly!

## 0.5.13
* Add note to README.md about linting on-the-fly

## 0.5.12
* Fix error with having a file open but no project.
* Add usage notes for ```-fsyntax-only``` and ```-c``` flags.

## 0.5.11
* Add Cross-platform ```$PATH``` expansion

## 0.5.10
* Removed compile flag to fix problem with GCC on OSX, moved option to config page

## 0.5.9
* Add compile flag to GCC command

## 0.5.8
* Add GCC to messages to play nice with linter-clang

## 0.5.7
* Add command-line error detection

## 0.5.6
* Beautify and add descriptions to package settings.

## 0.5.5
* Added unit tests for C source files

## 0.5.4
* Corrected error in package manifest

## 0.5.3
* Correctly atom-package-deps to manifest and main.js

## 0.5.2
* Removed atom-package-deps

## 0.5.1
* Added package-deps for automatic installation of base linter

## 0.5.0
* Added hierarchical configuration options
* Added unit tests for file/directory-specific settings

## 0.4.1
* Added build image to README.md

## 0.4.0
* Fixed leading space being added to message
* Added travis CI integration

## 0.3.9
* Fixed problem where fatal errors were ignored by the linter.
* Improved debugging output - full command printed directly.

## 0.3.8
* Fixed problem with parsing of multiple 'include paths' if some were empty.

## 0.3.7
* Fixed problem where an empty 'include paths' parameter prevented linting.

## 0.3.6
* Merge [Kepler](https://github.com/k2b6s9j)'s pull request to improve regex type capture

## 0.3.5
* Added descriptive keywords

## 0.3.4
* Updated description to be more search-friendly
* Re-ordered CHANGELOG.md so newest version is on top

## 0.3.3
* Fixed bug in getting file path

## 0.3.2
* Root folder is file path if no projects present

## 0.3.1
* Added relative path expansion to execPath parameter
* Added debug logging to console

## 0.3.0
* .gcc-flags.json now reloaded on file save
* Relative paths expanded with respect to project root

## 0.2.1
* Fixed .gcc-flags.json cache issue

## 0.2.0
* Added Screenshot to README
* Added project-specific settings

## 0.1.1
* Cleaned up settings taken from linter-clang

## 0.1.0 - First Release
* First Release - tested on Ubuntu x64, gcc and g++


spec/files/error.c
#include <stdio.h>

int main(int argc, char const *argv[]) {
    return i;
}


spec/files/error.f
      program main
      implicit none
      i=1
      end


spec/config-spec.js
"use babel";

describe('Configuration function tests', () => {
  const main = require('../lib/main')
  const utility = require('../lib/utility.js')
  var settings = require("../lib/config").settings

  beforeEach(() => {
    waitsForPromise(() => {
      main.messages={};
      main.last_linted_files=new Set([]);
      atom.config.set('linter-gcc.execPath', '/usr/bin/g++')
      atom.config.set('linter-gcc.gccDefaultCFlags', '-Wall')
      atom.config.set('linter-gcc.gccDefaultCppFlags', '-Wall -std=c++11')
      atom.config.set('linter-gcc.gccErrorLimit', 15)
      atom.config.set('linter-gcc.gccIncludePaths', ' ')
      atom.config.set('linter-gcc.gccISystemPaths', ' ')
      atom.config.set('linter-gcc.gccSuppressWarnings', true)
      atom.config.set('linter-gcc.gccRelintMessageSources', false)
      atom.config.set('linter-gcc.gcc7orGreater', false)
      atom.config.set('linter-gcc.gccLintOnTheFly', false)
      atom.config.set('linter-gcc.gccDebug', false)
      atom.config.set('linter-gcc.gccErrorString', 'error')
      atom.config.set('linter-gcc.gccWarningString', 'warning')
      atom.config.set('linter-gcc.gccNoteString', 'note')
      atom.config.set('linter-gcc.compileCommandsFile', ' ')
      atom.packages.activatePackage('language-c')
      return atom.packages.activatePackage('linter-gcc')
    })
  })

  it('Uses default settings when no config file is found', () => {
    waitsForPromise(() => {
      return atom.workspace.open(__dirname + '/files/comment.cpp').then(() => {
          var config = settings()
          expect(config.execPath).toEqual("/usr/bin/g++")
          expect(config.gccDefaultCFlags).toEqual("-Wall")
          expect(config.gccDefaultCppFlags).toEqual("-Wall -std=c++11")
          expect(config.gccErrorLimit).toEqual(15)
          expect(config.gccIncludePaths).toEqual(" ")
          expect(config.gccSuppressWarnings).toEqual(true)
      })
    })
  })

  it('Uses file-specific config file when it exists', () => {
    waitsForPromise(() => {
      return atom.workspace.open(__dirname + '/files/project_test/sub1/subsub1/file.cpp').then(() => {
          var config = settings()
          expect(config.execPath).toEqual("exec_file")
          expect(config.gccDefaultCFlags).toEqual("cflags_file")
          expect(config.gccDefaultCppFlags).toEqual("cppflags_file")
          expect(config.gccErrorLimit).toEqual(1)
          expect(config.gccIncludePaths).toEqual("includepath_file")
          expect(config.gccSuppressWarnings).toEqual(true)
      })
    })
  })

  it('Uses directory-specific config file when it exists', () => {
    waitsForPromise(() => {
      return atom.workspace.open(__dirname + '/files/project_test/sub2/file.cpp').then(() => {
          var config = settings()
          expect(config.execPath).toEqual("exec_subdir")
          expect(config.gccDefaultCFlags).toEqual("cflags_subdir")
          expect(config.gccDefaultCppFlags).toEqual("cppflags_subdir")
          expect(config.gccErrorLimit).toEqual(2)
          expect(config.gccIncludePaths).toEqual("includepath_subdir")
          expect(config.gccSuppressWarnings).toEqual(true)
      })
    })
  })

  it('Uses current directory config file when it exists', () => {
    waitsForPromise(() => {
      return atom.workspace.open(__dirname + '/files/project_test/sub2/file.cpp').then(() => {
          var config = settings()
          expect(config.execPath).toEqual("exec_subdir")
          expect(config.gccDefaultCFlags).toEqual("cflags_subdir")
          expect(config.gccDefaultCppFlags).toEqual("cppflags_subdir")
          expect(config.gccErrorLimit).toEqual(2)
          expect(config.gccIncludePaths).toEqual("includepath_subdir")
          expect(config.gccSuppressWarnings).toEqual(true)
      })
    })
  })

  it('Uses upper-level config file when it exists', () => {
    waitsForPromise(() => {
      atom.project.addPath(__dirname + '/files/project_test')
      return atom.workspace.open(__dirname + '/files/project_test/sub4/subsub2/file.cpp').then(() => {
          var config = settings()
          expect(config.execPath).toEqual("exec_updir")
          expect(config.gccDefaultCFlags).toEqual("cflags_updir")
          expect(config.gccDefaultCppFlags).toEqual("cppflags_updir")
          expect(config.gccErrorLimit).toEqual(5)
          expect(config.gccIncludePaths).toEqual("includepath_updir")
          expect(config.gccSuppressWarnings).toEqual(true)
      })
    })
  })

  it('Uses project-specific config file when it exists', () => {
    waitsForPromise(() => {
      atom.project.addPath(__dirname + '/files/project_test')
      return atom.workspace.open(__dirname + '/files/project_test/sub3/file.cpp').then( () => {
          var config = settings()
          expect(config.execPath).toEqual("exec_project")
          expect(config.gccDefaultCFlags).toEqual("cflags_project")
          expect(config.gccDefaultCppFlags).toEqual("cppflags_project")
          expect(config.gccErrorLimit).toEqual(3)
          expect(config.gccIncludePaths).toEqual("includepath_project")
          expect(config.gccSuppressWarnings).toEqual(false)
      })
    })
  })
})


lib/config.js
'use strict';

var path = require('path');
var fs = require('fs');
var config_file_cache = new Map();

module.exports.niceName = 'Custom file (.gcc-flags.json)';

module.exports.settings = function () {

  var SETTINGS_FILENAME = ".gcc-flags.json";
  var MAX_ITERATIONS = 30;

  var linted_file = atom.workspace.getActiveTextEditor().getPath();
  var file_settings = linted_file + SETTINGS_FILENAME;
  var directory_settings = path.join(path.dirname(file_settings), SETTINGS_FILENAME);
  var config_file = "";
  if (fs.existsSync(file_settings)) {
    config_file = file_settings;
  } else if (fs.existsSync(directory_settings)) {
    config_file = directory_settings;
  }

  let project_path = atom.project.relativizePath(linted_file)[0];
  if (project_path === null) {
    project_path = path.dirname(linted_file);
  }

  if (config_file == "") {
    var current_path = path.dirname(file_settings);
    var current_file = "";
    var counter = 0;
    while (path.relative(current_path, project_path) != "" && counter < MAX_ITERATIONS){
      current_path = path.join(current_path, "..");
      current_file = path.join(current_path, SETTINGS_FILENAME);
      if (fs.existsSync(current_file)) {
        config_file = current_file;
        break;
      }
      counter += 1;
    }
  }

  if (atom.config.get("linter-gcc.gccDebug")){
    if (config_file.length > 0) {
      console.log("linter-gcc: Reading settings from: " + config_file);
    } else {
      console.log("linter-gcc: Using configuration page settings");
    }
  }
  var commands_file = ""
  if (config_file != "") {
    var last_modified = fs.statSync(config_file).mtime.getTime();
    var config_data = {};
    if (config_file_cache.has(config_file) && config_file_cache.get(config_file).last_modified == last_modified)
    {
      config_data = config_file_cache.get(config_file).config_data;
    }
    else
    {
      config_data = JSON.parse(fs.readFileSync(config_file));
      config_file_cache.set(config_file, {last_modified: last_modified, config_data: config_data});
    }
    if ("compileCommandsFile" in config_data) {
      commands_file = config_data.compileCommandsFile;
    } else {
      commands_file = atom.config.get("linter-gcc.compileCommandsFile");
    }
    return {
      execPath: config_data.execPath,
      gccIncludePaths: config_data.gccIncludePaths,
      gccISystemPaths: config_data.gccISystemPaths,
      gccSuppressWarnings: config_data.gccSuppressWarnings,
      gcc7orGreater: config_data.gcc7orGreater,
      gccDefaultCFlags: config_data.gccDefaultCFlags,
      gccDefaultCppFlags: config_data.gccDefaultCppFlags,
      gccErrorLimit: config_data.gccErrorLimit,
      gccErrorString: config_data.gccErrorString,
      gccWarningString: config_data.gccWarningString,
      gccNoteString: config_data.gccNoteString,
      compileCommandsFile: commands_file
    };
  } else {
    return {
      execPath: atom.config.get("linter-gcc.execPath"),
      gccIncludePaths: atom.config.get("linter-gcc.gccIncludePaths"),
      gccISystemPaths: atom.config.get("linter-gcc.gccISystemPaths"),
      gccSuppressWarnings: atom.config.get("linter-gcc.gccSuppressWarnings"),
      gcc7orGreater: atom.config.get("linter-gcc.gcc7orGreater"),
      gccDefaultCFlags: atom.config.get("linter-gcc.gccDefaultCFlags"),
      gccDefaultCppFlags: atom.config.get("linter-gcc.gccDefaultCppFlags"),
      gccErrorLimit: atom.config.get("linter-gcc.gccErrorLimit"),
      gccErrorString: atom.config.get("linter-gcc.gccErrorString"),
      gccWarningString: atom.config.get("linter-gcc.gccWarningString"),
      gccNoteString: atom.config.get("linter-gcc.gccNoteString"),
      compileCommandsFile: atom.config.get("linter-gcc.compileCommandsFile"),
    };
  }

};


lib/gccMessage.js
class GccMessage {
  constructor(colStart) {
    this.colStart = colStart;
  }
  colEnd() {
    return this.colStart + 1;
  }
}

function extractVariableName(message) {
  var match1 =  message.match(/\u2018.*\u2019/)[0];
  return match1.replace(/\u2018|\u2019/g, '');
}

class UndeclaredError extends GccMessage {
  constructor(colStart, message){
    super(colStart);
    this.varName = extractVariableName(message);
  }
  colEnd() {
    return this.colStart + this.varName.length;
  }
}

class UnusedVariableWarning extends GccMessage {
  constructor(colStart, message){
    super(colStart);
    this.varName = extractVariableName(message);
  }
  colEnd() {
    return this.colStart + this.varName.length;
  }
}

module.exports = {
  build: function(colStart, message) {
    if (message.includes("was not declared in this scope")) {
      return new UndeclaredError(colStart, message);
    }
    else if (message.includes("unused variable")) {
      return new UnusedVariableWarning(colStart, message);
    }
    else {
      return new GccMessage(colStart);
    }
  }
}


lib/main.js
'use babel'

var time_last_lint = new Date().getTime();
var lint_waiting = false;

CompositeDisposable = require('atom').CompositeDisposable;

module.exports = {
  config: {
    execPath: {
      title: "GCC Executable Path",
      description: "Note for Windows/Mac OS X users: please ensure that GCC is in your ```$PATH``` otherwise the linter might not work. If your path contains spaces, it needs to be enclosed in double quotes.",
      type: "string",
      default: "g++",
      order : 1
    },
    gccDefaultCFlags: {
      title: "C Flags",
      description: "Supports the use of escaped characters",
      type: "string",
      default: "-c -Wall",
      order : 2
    },
    gccDefaultCppFlags: {
      title: "C++ Flags",
      description: "Supports the use of escaped characters",
      type: "string",
      default: "-c -Wall -std=c++11",
      order : 3
    },
    gccIncludePaths: {
      title: "GCC Include Paths",
      description: "Enter your include paths as a comma-separated list. Paths starting with ```.``` or ```..``` are expanded relative to the project root path and paths starting with a ```-``` are expanded relative to the path of the active file. If any of your paths contain spaces, they need to be enclosed in double quotes. To expand a directory recursively, add ```/*``` to the end of the path",
      type: "string",
      default: " ",
      order : 4
    },
    gccISystemPaths: {
      title: "GCC -isystem Include Paths",
      description: "Enter your include paths the same as above. These will be included using the -isystem option, treating the included files like system files and ignoring their warnings.",
      type: "string",
      default: " ",
      order : 5
    },
    gccErrorLimit: {
      title: "GCC Error Limit",
      description: "To completely remove `-fmax-errors`, simply enter `-1` here.",
      type: "integer",
      default: 0,
      order : 6
    },
    gccSuppressWarnings: {
      title: "Suppress GCC Warnings",
      type: "boolean",
      default: false,
      order : 7
    },
    gccErrorString: {
        title: "String GCC prepends to errors",
        type: "string",
        default: "error",
        order: 8
    },
    gccWarningString: {
        title: "String GCC prepends to warnings",
        type: "string",
        default: "warning",
        order: 9
    },
    gccNoteString: {
        title: "String GCC prepends to notes",
        type: "string",
        default: "note",
        order: 10
    },
    gccLintOnTheFly: {
      title: "Lint on-the-fly",
      description: "Please ensure any auto-saving packages are disabled before using this feature",
      type: "boolean",
      default: false,
      order : 11
    },
    gccLintOnTheFlyInterval: {
      title: "Lint on-the-fly Interval",
      description: "Time interval (in ms) between linting",
      type: "integer",
      default: 300,
      order : 12
    },
    gccRelintMessageSources: {
      title: "Re-lint message source files",
      description: "Reruns linter on files that contributed messages to the current working file",
      type: "boolean",
      default: false,
      order : 13
    },
    gccDebug: {
      title: "Show Debugging Messages",
      description: "Please read the linter-gcc wiki [here](https://github.com/hebaishi/linter-gcc/wiki) before reporting any issues.",
      type: "boolean",
      default: false,
      order : 14
    },
    gcc7orGreater: {
      title: "Gcc version 7 or newer",
      description: "Check if you have a Gcc version that recognizes \"-fdiagnostics-parseable-fixits\" (7 or newer)",
      type: "boolean",
      default: true,
      order : 15
    },
    compileCommandsFile: {
      title: "Compile commands file",
      description: "Path to cmake compile_commands.json",
      type: "string",
      default: "./build/compile_commands.json",
      order : 16
    }
  },

  messages: {},
  last_linted_files: {},
  linter_gcc: undefined,

  temp_file : {
    "C++": require("tempfile")(".cpp"),
    "C": require("tempfile")(".c")
  },

  lint: function(editor, linted_file, real_file){
    //create list of extra files to lint
    otherFiles = new Set([])
    if (atom.config.get("linter-gcc.gccRelintMessageSources") == true){
      for (var file in module.exports.messages){
        if (!module.exports.messages.hasOwnProperty(file)) continue;
        var fileMessages = module.exports.messages[file];
        for (var msg in fileMessages){
          if (!fileMessages.hasOwnProperty(msg)) continue;
          var message = fileMessages[msg];
          if (message.location.file === real_file && file !== real_file && !module.exports.last_linted_files.has(file)){
            otherFiles.add(file);
          }
        }
      }
      for (let file of otherFiles){
        module.exports.lint(editor, file, file);
      }
    }
    module.exports.last_linted_files.add(real_file)
    command = require("./utility").buildCommand(editor, linted_file, real_file);
    return require("sb-exec").exec(command.binary, command.args, {stream: "stderr"}).then(output => {
      msgs = require("./utility.js").parse(output,editor)
      msgs.forEach(function(entry){
        if (entry.location.file === module.exports.temp_file["C"] || entry.location.file === module.exports.temp_file["C++"]){
          entry.location.file = real_file;
        }
        if (entry.severity ===`${atom.config.get("linter-gcc.gccWarningString")}`)
            entry.severity = "warning";
        else if (entry.severity ===`${atom.config.get("linter-gcc.gccErrorString")}`)
            entry.severity = "error";
        else if (entry.severity ===`${atom.config.get("linter-gcc.gccNoteString")}`)
            entry.severity = "info";
      })

      //If parsing failed, return a general message if there was an error
      if (msgs.length == 0 && output.indexOf("error") != -1){
        msgs = [{
          severity: 'error',
          excerpt: output,
          location: {
              file: real_file,
              position: [[0, 0], [0, 0]]
          }
        }];
      }
      module.exports.messages[real_file] = msgs;
      // console.log(msgs)
      if (typeof module.exports.linter_gcc != "undefined"){
        module.exports.linter_gcc.setAllMessages(require("./utility").flattenHash(module.exports.messages))
      }
      return msgs;
    })
  },

  activate: function() {
    this.subscriptions = new CompositeDisposable()
    if(!atom.packages.getLoadedPackages("linter")) {
    atom.notifications.addError(
      "Linter package not found.",
      {
        detail: "Please install the `linter` package in your Settings view."
      }
    );
    }
    require("atom-package-deps").install("linter-gcc");
    time_last_lint = new Date().getTime()
    lint_waiting = false
  },
  deactivate: function() {
    this.subscriptions.dispose()
  },
  consumeLinter: function(registerIndie) {
    module.exports.linter_gcc = registerIndie({
      name: 'GCC'
    })

    subs = this.subscriptions;
    utility = require("./utility")
    lintOnTheFly = function() {
      module.exports.last_linted_files = new Set([])
      editor = utility.getValidEditor(atom.workspace.getActiveTextEditor());
      if (!editor) return;
      if (atom.config.get("linter-gcc.gccLintOnTheFly") == false) return;
      if (lint_waiting) return;
      lint_waiting = true
      interval = atom.config.get("linter-gcc.gccLintOnTheFlyInterval")
      time_now = new Date().getTime()
      timeout = interval - (time_now - time_last_lint);
      setTimeout(
        function() {
          time_last_lint = new Date().getTime()
          lint_waiting = false
          grammar_type = utility.grammarType(editor.getGrammar().name)
          filename = String(module.exports.temp_file[grammar_type])
          require('fs-extra').outputFileSync(filename, editor.getText());
          module.exports.lint(editor, filename, editor.getPath());
        },
        timeout
      );
    };

    lintOnSave = function(){
      module.exports.last_linted_files = new Set([])
      editor = utility.getValidEditor(atom.workspace.getActiveTextEditor());
      if (!editor) return;
      if (atom.config.get("linter-gcc.gccLintOnTheFly") == true) return;
      real_file = editor.getPath();
      //cleanup messages in this file created by linting other files
      if (atom.config.get("linter-gcc.gccRelintMessageSources") == false){
        for (var file in module.exports.messages){
          if (!module.exports.messages.hasOwnProperty(file)) continue;
          var fileMessages = module.exports.messages[file];
          for (var msg in fileMessages){
            if (!fileMessages.hasOwnProperty(msg)) continue;
            var message = fileMessages[msg];
            if (message.location.file === real_file){
              delete fileMessages[msg];
            }
          }
        }
      }
      module.exports.lint(editor, real_file, real_file);
    };

    cleanupMessages = function(){
      editor_hash = {};
      atom.workspace.getTextEditors().forEach( function(entry){
        try{
          path = entry.getPath()
        } catch(err){
        }
        editor_hash[entry.getPath()] = 1;
      });
      for (var file in module.exports.messages){
        if (!editor_hash.hasOwnProperty(file)){
          delete module.exports.messages[file]
        }
      }
      module.exports.linter_gcc.setAllMessages(JSON.parse(JSON.stringify(require("./utility").flattenHash(module.exports.messages))));
    };

    subs.add(module.exports.linter_gcc)

    atom.workspace.observeTextEditors(function(editor) {
      subs.add(editor.onDidSave(lintOnSave))
      subs.add(editor.onDidStopChanging(lintOnTheFly))
      subs.add(editor.onDidDestroy(cleanupMessages))
    })
  }
}
